metricbeat/module/elasticsearch/index/data_xpack.go

```diff
@@ -0,0 +1,332 @@
+			"norms_memory_in_bytes":         c.Int("norms_memory_in_bytes"),
+			"points_memory_in_bytes":        c.Int("points_memory_in_bytes"),
+			"doc_values_memory_in_bytes":    c.Int("doc_values_memory_in_bytes"),
+			"index_writer_memory_in_bytes":  c.Int("index_writer_memory_in_bytes"),
+			"version_map_memory_in_bytes":   c.Int("version_map_memory_in_bytes"),
+			"fixed_bit_set_memory_in_bytes": c.Int("fixed_bit_set_memory_in_bytes"),
+		}),
+		"store": c.Dict("store", s.Schema{
+			"size_in_bytes": c.Int("size_in_bytes"),
+		}),
+		"refresh": c.Dict("refresh", s.Schema{
+			"total_time_in_millis": c.Int("total_time_in_millis"),
+		}),
+	}
+
+	cacheStatsSchema = s.Schema{
+		"memory_size_in_bytes": c.Int("memory_size_in_bytes"),
+		"evictions":            c.Int("evictions"),
+		"hit_count":            c.Int("hit_count"),
+		"miss_count":           c.Int("miss_count"),
+	}
+)
+
+func eventsMappingXPack(r mb.ReporterV2, m *MetricSet, info elasticsearch.Info, content []byte) error {
+	err := json.Unmarshal(content, &indicesStruct)
+	if err != nil {
+		return err
+	}
+
+	clusterStateMetrics := []string{"metadata", "routing_table"}
+	clusterState, err := elasticsearch.GetClusterState(m.HTTP, m.HTTP.GetURI(), clusterStateMetrics)
+	if err != nil {
+		return err
+	}
+
+	for name, index := range indicesStruct.Indices {
+		event := mb.Event{}
+		indexStats, err := xpackSchema.Apply(index)
+		if err != nil {
+			continue
+		}
+		indexStats["index"] = name
+
+		err = addClusterStateFields(name, indexStats, clusterState)
+		if err != nil {
+			continue
+		}
+
+		event.RootFields = common.MapStr{
+			"cluster_uuid": info.ClusterID,
+			"timestamp":    common.Time(time.Now()),
+			"interval_ms":  m.Module().Config().Period / time.Millisecond,
+			"type":         "index_stats",
+			"index_stats":  indexStats,
+		}
+
+		event.Index = elastic.MakeXPackMonitoringIndexName(elastic.Elasticsearch)
+		r.Event(event)
+	}
+
+	return nil
+}
+
+func addClusterStateFields(indexName string, indexStats, clusterState common.MapStr) error {
+	indexMetadata, err := getClusterStateMetricForIndex(clusterState, indexName, "metadata")
+	if err != nil {
+		return err
+	}
+
+	indexRoutingTable, err := getClusterStateMetricForIndex(clusterState, indexName, "routing_table")
+	if err != nil {
+		return err
+	}
+
+	// TODO:
+
+	// "index_stats.created",
+	v, err := indexMetadata.GetValue("settings.index.creation_date")
+	if err != nil {
+		return err
+	}
+	c, ok := v.(string)
+	if !ok {
+		return elastic.MakeErrorForMissingField("settings.index.creation_date", elastic.Elasticsearch)
+	}
+
+	created, err := strconv.ParseInt(c, 10, 64)
+	if err != nil {
+		return err
+	}
+	indexStats.Put("created", created)
+
+	// "index_stats.version.created", <--- don't think this is being used in the UI, so can we skip it?
+	// "index_stats.version.upgraded", <--- don't think this is being used in the UI, so can we skip it?
+
+	// "index_stats.status",
```

Comment:
This comment may indicate that the following lines could be in its own funcion, as you do with `getIndexStatus()` and `getIndexShardsStats()` (even if here it'd be a smaller function). The same for `index_stats.created` before.